File: DSA/Typescript/LeetCode.TypeScript/3sum.ts
export function threeSum(nums: number[]): number[][] {
    nums.sort((a,b) => a - b);
    let resultingArray: number[][] = [];
    for (let i = 0; i < nums.length - 2; i++) {
        if(i > 0 && nums[i] === nums[i-1]){
            continue;
        }
        for (let j = nums.length-1; j >= i+2; j--) {
            if(j < nums.length-1 && nums[j] === nums[j+1]){
                continue;
            }

            let difference = nums[i] + nums[j];
            let indexOfMiddleOperand = -1;
            if(difference <= 0){
                for( let m = j-1; m >= i+1; m--){
                    if(nums[m] + difference < 0){
                        break;
                    }
                    if(nums[m] + difference == 0){
                        indexOfMiddleOperand = m;
                        break;
                    }
                    
                }
            } else {
                for( let m = i+1;  m <= j-1; m++){
                    if(nums[m] + difference > 0){
                        break;
                    }
                    if(nums[m] + difference == 0){
                        indexOfMiddleOperand = m;
                        break;
                    }
                    
                }
            }
            if(indexOfMiddleOperand !== -1){
                resultingArray.push([nums[i], nums[indexOfMiddleOperand], nums[j]]);
            }
      
        }
    }

    return resultingArray;
};

File: DSA/Dotnet/LeetCode.Net/Problems/Two Pointers/3sum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime;
using System.Runtime.CompilerServices;

namespace LeetCode.Problems.TwoPointers;

public class ThreeSumSolution
{
    public List<List<int>> ThreeSum(int[] nums)
    {
        var pairs = new List<List<int>>();
        Array.Sort(nums);

        for (var t = 0; t < nums.Length - 2; t++)
        {
            if (t > 0 && nums[t] == nums[t - 1]) continue;

            var left = t + 1;
            var right = nums.Length - 1;

            while (left < right)
            {
                var sum = nums[t] + nums[left] + nums[right];
                if (sum == 0)
                {
                    pairs.Add(new List<int>() { nums[t], nums[left], nums[right] });

                    while (left < right && nums[left] == nums[left + 1]) left++;
                    while (left < right && nums[right] == nums[right - 1]) right--;

                    left++;
                    right--;
                }
                else if (sum < 0)
                {
                    left++;
                }
                else
                {
                    right--;
                }
            }
        }

        return pairs;
    }
}
